[CursorInfo] Implement a few expression references as solver-based

ahoppen opened the pull request:
This implements cursor info resolving for a few expression types using the constraint system. This allows us to detect ambiguous results – we cannot deliver them yet but that will be done in a follow-up PR.

ahoppen committed:
[CursorInfo] Implement a few expression references as solver-based

This implements cursor info resolving for a few expression types using the constraint system. This allows us to detect ambiguous results – we cannot deliver them yet but that will be done in a follow-up PR.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
Stress tester failure at `SceneDelegate.swift:unmodified:1981` is an improvement. A reduced reproducer is
```swift
protocol MyProto {
    static func makeInstance() -> Self
}

class Base: MyProto {
    static func makeInstance() -> Self {
        fatalError()
    }
}

class Inherited: Base {
    var member: Int
}

// RUN: %sourcekitd-test -req=cursor -pos=%(line + 2):30 %s -- %s
func foo() {
    Inherited.makeInstance().member = 2
}
```

The AST-based implementation reported `Base` as the receiver type, the solver-based implementation reports `Inherited`, which is correct.

ahoppen commented:
Stress tester failure at `SceneDelegate.swift:concurrent-708:703` is an improvement. 

The AST-based implementation reported `shared` as a `class` variable, but it’s spelled as a `static` variable in source.

ahoppen commented:
Stress tester failure at `APIService.swift:unmodified:3766` is an improvement.

In the AST-based implementation on `async` in `DispatchQueue.main.async` no receiver was reported, the solver-based implementation reports `OS_dispatch_queue`.

ahoppen commented:
@swift-ci Please SourceKit stress test

bnbarham commented:
> The AST-based implementation reported Base as the receiver type, the solver-based implementation reports Inherited, which is correct.

I wouldn't expect any receiver type here, it's statically resolvable. If the old implementation is returning dynamic here I'd say that's a bug.

> In the AST-based implementation on async in DispatchQueue.main.async no receiver was reported, the solver-based implementation reports OS_dispatch_queue.

Is `DispatchQueue.async` `@objc`? If it isn't then as it's defined in an extension, it isn't overridable, and thus shouldn't be dynamic or have a receiver. As far as I can tell, it isn't, though maybe we're just missing the annotation in the generated interface.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented on lib/IDE/CursorInfo.cpp:
```diff
@@ -242,6 +331,59 @@ class CursorInfoDoneParsingCallback : public IDEInspectionCallbacks {
+      return nullptr;
+    }
+
+    // Deliver results
```
All the results are reported using callback functions because they have pointers to stack variables – which they really shouldn’t have but that’s something else I want to fix. Thus, we cannot store the results, run the AST-based cursor info and then report them. Hence, what we do right now is serialize the results into a string and just deliver the AST-based results.

So: Delivering the results is a non-trivial change that I want to do in a follow-up PR, especially since we want to fall back to AST-based cursor info if solver-based cursor info doesn’t return any results.

bnbarham replied:
To be clear, we're now only delivering the one result right 😅?

ahoppen replied:
Yes, for now. Fixing the `TODO` in the line above is the next thing to do.

```cpp
// FIXME: We need to be able to report multiple results.
```

ahoppen committed:
[CursorInfo] Deliver results from solver-based cursor info

Running the SourceKit stress tester with verification of solver-based cursor info returned quite a few differences but in all of them, the old AST-based implementation was actually incorrect. So, instead of verifying  the results, deliver the results from solver-baesd cursor info and only fall back to AST-based cursor info if the solver-based implementation returned no results.

rdar://103369449

ahoppen committed:
[CursorInfo] A few miscellaneous fixes to the AST-based cursor info and new test cases

In these cases the solver-based and AST-based cursor info differed in their results. Fix the AST-based cursor info to return the correct results and add test cases to make sure the solver-based implementation doesn’t regress them.

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen committed:
[CursorInfo] Fix bug when performing solver-based cursor info on Self

ahoppen committed:
[CursorInfo] Fix a bug that caused solver-based cursor info to crash if invoked on a symbol from a different module

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
Note to self: The stress tester only found failures in refactoring where previously we didn’t report any cursor, so all of those are improvements.

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/AST/Decl.cpp:
```diff
@@ -1829,8 +1829,9 @@ SourceRange PatternBindingDecl::getSourceRange() const {
 static StaticSpellingKind getCorrectStaticSpellingForDecl(const Decl *D) {
   if (auto classDecl = D->getDeclContext()->getSelfClassDecl()) {
-    if (!classDecl->isActor())
+    if (!classDecl->isActor() && !isa<VarDecl>(D)) {
       return StaticSpellingKind::KeywordClass;
+    }
```
Re: `!isa<VarDecl>(D)`, does this affect `class` computed values? e.g. 
```swift
class C {
  class var value: Int { 1 }
}
```

ahoppen commented:
@swift-ci Please smoke test

rintaro commented on lib/IDE/Utils.cpp:
```diff
@@ -956,13 +961,13 @@ void swift::ide::getReceiverType(Expr *Base,
+  if (auto SelfT = ReceiverTy->getAs<DynamicSelfType>())
     ReceiverTy = SelfT->getSelfType();
 
+  ReceiverTy->getWithoutSpecifierType();
```
Forgot to remove?

ahoppen commented:
@swift-ci Please smoke test

ahoppen commented:
@swift-ci Please SourceKit stress test

ahoppen commented:
@swift-ci Please smoke test Windows
